### PR TITLE
VACMS-8141 Prevent re-use of VAMC banner alerts.

### LINKED ISSUE #8141: VAMC editor are re-using VAMC System Banner Alert with Situation Updates
## User Story or Problem Statement

Editors should not be  able to re-use an alert.  

Possible solutions:
A) When they unselect "published", the node gets deleted  (nuclear option... not ideal)
B) When they unselect "published",  that checkbox becomes disabled and replaced with a message that alerts can not be re-enabled.
C) When they unselect "published", edit is not possible (remove edit tab for non-admin)
D) When they unselect "published", save of edit is not possible (remove save button for non-admin)

## Acceptance Criteria
- [x] I expect after editors save a VAMC system banner alert with the checkbox published "unselected" THEN edit is not possible (remove edit tab for non-admin)
- [x] I expect there is a message on the screen for editors that persists, work with @BlakeOrgan and @thejordanwood  to craft exact copy for on screen message (or other designated CMS copy stakeholder)
- [x] I expect if the VAMC system banner alert has never been published THEN it can be published 
- [x] I expect screenshots are sent to @kmariepat-cityfriends for KB article update 
- [ ] I expect PR is not merged to main, but instead merged to integration branch as part of #14117 (aka: ticket can close without merged PR)

## Implementation steps
<optional>

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team
Please check the team(s) that will do this work. 

- [ ] `CMS Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide CMS Team ` (leave Sitewide unchecked and check the specific team instead)
  - [ ] `⭐️ Content ops`
  - [ ] `⭐️ CMS experience`
  - [ ] `⭐️ Offices`
  - [ ] `⭐️ Product support`
  - [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by swirtSJW**
## Description

Closes #8141

## QA steps

Login as an admin
1 visit https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/admin/reports/status/rebuild  and rebuild the permissions (get a cup o coffee, this takes a while)

Login as user edgar.cruse  (an editor in chillicothe)
### Sections page
2. Go [here](https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/section/vha/vamc-facilities/va-chillicothe-health-care) and filter the page for content type = "VAMC System Banner Alert with Situation Updates"
   - [ ] Validate that some have edit buttons and others do not (indicating that some have already been published and unpublished and are no longer available to edit)
 3. GO to the content page https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/admin/content?title=&type=full_width_banner_alert&moderation_state=All&owner=337
   - [ ] Validate that the list has some with edit buttons and some without.
 Go to content not  in your section (tampa health care) https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/admin/content?title=&type=full_width_banner_alert&moderation_state=All&owner=319
    - [ ] Validate that none have edit buttons (Edgar doesn't belong here)
### Edit an existing published
3. go to [this published banner](https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/va-chillicothe-health-care/vamc-banner-alert/2021-11-16/coronavirus)  
   - [ ] validate that you can see the edit and revisions tabs.
   - [ ] Click edit.  Make no changes to the node except for adding a revision log. Save the node.  Validate that you can again see an Edit and Revisions tab
5. Click edit.  Make and uncheck the Published box, add a revision log. Save the node.
   - [ ] Validate that you NOT see an Edit tab, but the Revisions tab is visible.
   - [ ] Validate that you see a warning (yellow) indicating the node has been unpublished and can't be re-used.
  
image

6. Reload the page
   - [ ] Validate the warning persists and the edit tab is still not present.

### A banner that does not belong to you
1. Go to https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/va-tampa-health-care/vamc-banner-alert/2023-04-05/toxic-exposure-screening
   - [ ] Validate you have no access to edit or see revisions.
2. Visit an unpublished banner alert  https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/vamc-banner-alert/2021-06-22/website-coming-soon-not-the-official-va-tampa-health-care-website
   - [ ] Validate you have no way to edit and that you see a warning indicating it can't be edited.

### A new banner

1. Go to https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/node/add/full_width_banner_alert 
2. Fill out all required fields
3. Leave 'Published' unchecked.
4. Save the node
   - [ ] Validate that you still see an edit tab.
5. Edit the node and publish it.   
   - [ ] Validate that you still see an edit  and revisions tab. 
6. Edit the node, unpublish and save it
   - [ ] Validate that you do not, see an edit tab
   - [ ] Validate you do see a revisions tab
   - [ ] Validate you see a warning that says the node as been unpublished and is no longer editable.
 
Login as admin
### Sections page
2. Go [here](https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/section/vha/vamc-facilities/va-chillicothe-health-care) and filter the page for content type = "VAMC System Banner Alert with Situation Updates"
   - [ ] Validate that you see edit buttons on all the banners.
 3. go to this unpubleished previousley published banner https://pr14181-rfsw4ifjca4fum4i40ebyiux9avcuvkg.ci.cms.va.gov/va-chillicothe-health-care/vamc-banner-alert/2021-11-16/coronavirus
   - [ ] Validate that you can edit if you have to
   - [ ] Validate that you see a warning that you can not edit it.  (this is only going to appear once per cached version.)

### CODE REVIEW

**Comment by swirtSJW** [comment]
Just noting that the existing tests are all passing, before I add the cypress test.   No functional code should need chaning, just tweaking of the test 
image

**Comment by swirtSJW** [comment]
@omahane I just updated the cypress test so that it is passing (fingers crossed).  You should be good to review this.  i am leaving it as draft as it will have to wait until launch.

**Comment by swirtSJW** [comment]
@ndouglas Adding you as a reviewer because I added some new cypress steps.

File: tests/cypress/integration/common/should_exist.js [hunk 1]
```diff
@@ -34,3 +34,11 @@ Then("an image with the selector {string} should exist", (selector) => {
       expect($img[0].naturalHeight).to.be.greaterThan(0);
     });
 });
+
+Then("the primary tab {string} should exist", (text) =>
+  cy.get(".tabs__tab a").contains(text).should("exist")
```
Comment (ndouglas) [inline tests/cypress/integration/common/should_exist.js:39]: **Nitpick**: Should end with a semicolon. Wondering if I've broken ESLint 🤔

File: tests/cypress/integration/common/should_exist.js [hunk 2]
```diff
@@ -34,3 +34,11 @@ Then("an image with the selector {string} should exist", (selector) => {
...
+);
+
+Then("the primary tab {string} should not exist", (text) =>
+  cy.get(".tabs__tab a").contains(text).should("not.exist")
```
Comment (ndouglas) [inline tests/cypress/integration/common/should_exist.js:43]: **Nitpick**: Should end with a semicolon. Wondering if I've broken ESLint 🤔

**Review by ndouglas — COMMENTED** [review]
This looks fantastic to me. A couple of nitpicks for code style but that's it.

File: tests/cypress/integration/content_editing_vamc_banner_reuse.feature
Comment (ndouglas) [inline tests/cypress/integration/content_editing_vamc_banner_reuse.feature:1]: This is a really good test and I think the new step definitions will be very handy.

**Comment by ndouglas** [comment]
Cypress test failure unrelated.

I believe the test failure is because the current alert re: staff profiles is pushing the download link down to behind the JSD popup. It should really be testing for existence and not visibility. That's a test bug, I'll fix it.

Screenshot 2023-06-30 at 10 14 34 AM

**Comment by swirtSJW** [comment]
Looks like the failing test was not the banner re-use  it was media.   Seems unrelated.
image

```
  Media entities
    ✓ Log in and confirm that field_duration has h:m:s format (14162ms)
    ✓ Log in and create a document media entity. (25725ms)
    ✓ Log in and create an external document media entity. (25994ms)
    ✓ Log in and create an image media entity. (29012ms)
    ✓ Log in and create a video media entity. (28528ms)
    ✓ Log in and create a document media entity, a knowledge base article that refers to it, and verify a downloadable link is created. (64416ms)
    ✓ Log in and create an image media entity, a knowledge base page that refers to it, and verify a downloadable link is created. (67267ms)
    1) Log in and create a video media entity, a knowledge base page that refers to it, and verify a link is created.


  7 passing (6m)
  1 failing

  1) Media entities
       Log in and create a video media entity, a knowledge base page that refers to it, and verify a link is created.:
     AssertionError: Timed out retrying after 10000ms: Expected to find element: `//a[contains(@class, "downloadable-file-link--video") and contains(@target, "_blank")]`, but never found it.
      at Context.eval (tests/cypress/integration/common/i_should_see.js:61:4)
      at Context.resolveAndRunStepDefinition (node_modules/cypress-cucumber-preprocessor/lib/resolveStepDefinition.js:193:0)
      at Context.eval (node_modules/cypress-cucumber-preprocessor/lib/createTestFromScenario.js:27:0)

```

I'll make the suggested changes and try this again.

**Comment by swirtSJW** [comment]
Oops, just saw Nate's note about the failure.

File: tests/cypress/integration/common/should_exist.js (hunk 1 above)
Comment (swirtSJW) [reply tests/cypress/integration/common/should_exist.js:39]: It hollers at me if I put one in 
image

Comment (swirtSJW) [reply tests/cypress/integration/common/should_exist.js:39]: It looks like this format makes it happy.  

image

I can use the {} if you prefer, but that does not match all the other existing uses in the file.

Comment (ndouglas) [reply tests/cypress/integration/common/should_exist.js:39]: no no no, that was me misreading it and mentally throwing those curly braces in. My bad, everything works as intended, I'm the idiot.

**Comment by swirtSJW** [comment]
This PR just keeps getting stuck.  It builds and tests run and pass but they are not getting reported.  I am not sure if the build is not using the right commit hash or the reporting is just not happening.

**Comment by swirtSJW** [comment]
I just changed this to point to our intergration branch for banners, so if this passes @omahane  QA we are free to merge.

**Comment by swirtSJW** [comment]
Since this one takes a fair amount of testing, and we have to re-test the other two PRs in the integration branch,  I am going to merge this to integration, then we can run all tests there.

**Pull request merged**